### PR TITLE
Remove `is_stateless` flag

### LINKED ISSUE #268: Remove `is_stateless` flag
### What is wrong?

There's an `is_stateless` in `VM` configuration now: 
https://github.com/ethereum/py-evm/blob/4f476908dff42e290c15fc4cfb6a3c5e65e37dfe/evm/vm/base.py#L98-L127

In `is_stateless` mode, the `block` will be deeply copied and the get the result `block` from `VMState` with updated header and transactions. And in the non-stateless mode (the old route), the state transition result would be updated in `VM`.

### How can it be fixed

1. Remove the `is_stateless is False` route.
2. Modify `Chain.apply_transaction` to update the local vm state before returning the `computation` object.

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?
Fix #268 

### How was it fixed?
- [x] Removed `is_stateless` flag, only use the new route in `VM.apply_transaction`.

#### Cute Animal Picture

koala-630117_1280

### CODE REVIEW

File: evm/vm_state.py [hunk 1]
```diff
@@ -179,35 +179,27 @@ def get_computation(self, message):
...
-            # Set block.
-            block, trie_data = self.add_transaction(transaction, computation, block)
-            block.header.state_root = self.state_root
-            return computation, block, trie_data
-        else:
-            computation = self.execute_transaction(transaction)
-            block.header.state_root = self.state_root
-            return computation, None, None
+        # Don't modify the given block
+        block = copy.deepcopy(block)
```
Comment (pipermerriam) [inline evm/vm_state.py]: What do you think about removing this `deepcopy` and instead making `block` immutable and modifying the logic inside of `add_transaction` that creates and returns a new block object.

**Comment by hwwhww** [comment]
#### Changelogs
1. Updated Chain.header in Chain.apply_transaction + Added `chaindb.persist_trie_data_dict_to_db`
2. Create a pure new block in `VMState.add_transaction`

I added `BlockHeader.pure_clone(header)` class method to do the dumb copy:
[py-evm/headers.py at 8f4aab7eba7a17e9491c7bb49282a5ba7261670b · hwwhww/py-evm · GitHub](https://github.com/hwwhww/py-evm/blob/8f4aab7eba7a17e9491c7bb49282a5ba7261670b/evm/rlp/headers.py#L137-L155)

* An alternative solution is do the `copy.deepcopy(header)` in `VMState.add_transaction` and then call `rlp.make_mutable`, but it’s against the goal of removing `deepcopy`.
* I'm not sure whether the naming of `pure_clone` is appropriate, it's just my instinct.

File: evm/db/chain.py [hunk 2]
```diff
@@ -316,6 +316,13 @@ def persist_block_to_db(self, block):
             rlp.encode(block.uncles, sedes=rlp.sedes.CountableList(type(block.header))),
         )
 
+    def persist_trie_data_dict_to_db(self, trie_data_dict):
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply evm/db/chain.py]: But it _does_ use `self.db`...?

File: evm/rlp/headers.py [hunk 3]
```diff
@@ -133,3 +133,23 @@ def from_parent(cls,
 
         header = cls(**header_kwargs)
         return header
+
+    @classmethod
+    def pure_clone(cls, header):
+        return cls(
```
[the remark this answers is not in the document]
Comment (hwwhww) [reply evm/rlp/headers.py]: Your way looks much better for maintaining the possible new fields, thank you!

File: evm/db/chain.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hwwhww) [reply evm/db/chain.py]: But it _does_ use `self.db`...?

[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/db/chain.py]: Disregard, I missed it somehow

**Comment by hwwhww** [comment]
@pipermerriam Thank you for the review feedback, I added the patch commit.

**Pull request merged**